Service Virtualization Feature enhancement

elans3 opened the pull request:
I am one of the swagger/Open API user and would like to be a future contributor to the community. I had started developing a virtualized service idea and donate the code to Swagger community from the beginning. 

How can we use codegen project into Service Virtualization, As it holds tremendous potential to help testers boost test automation and keep pace with Agile and DevOps demands. If Developer/Tester being pressured to test faster and/or earlier in each release cycle (e.g., “shift left”), Service Virtualization could help advance development team’s Continuous Testing and development efforts and lot more in the API Contract driven world"

I have been working on project and had a requirement to develop a virtualized service. I had customized swagger codegen to create a virtualized service generation for the any given REST Yaml contract with JSON support.

How to setup virtualized service for all your Rest contracts in less then 10 minutes...
Required Info:
https://github.com/elans3/openapi-generator/blob/master/modules/openapi-generator-virtualan/src/main/resources/java-config.json

Setup Instruction: (All steps are same used as Swagger codegen as reference)
https://virtualmockservice.blogspot.com/2018/04/hi-everyone-in-this-microservice-world.html

Please let me know if you need any additional details. I would be eager to help integrate with codegen project.
Thanks,
Elan

jmini commented:
Sorry to ask stupid question, but how does this project is related to `openapi-generator`? I do not see where `openapi-generator` is used. What is the dependencies between the "Service Virtualization Feature" and "OpenAPI Generator"...

@wing328: maybe we could publish it as a new project inside the https://github.com/OpenAPITools organization?

@elans3 What do you think about this proposition? Can you give me more details?

elans3 commented:
@jmini 

Reason would be - Virtualized service had developed using the Open API Code generation project with  openapi-generator-cli  CLI tool.

Refer the codegen files that would generate as  "Virtualized" service.
- https://github.com/elans3/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/JavaSpring/apiControllerMockLogic.mustache
- https://github.com/elans3/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/JavaSpring/apiMockLogic.mustache
During code generation need to set choice for Virtualized service and would help create the Virtualized service of all Resources/YAML used to generate. 
                  -  "virtual-service": true (java-config.json) -  Currently it works only with spring
                  - java -jar openapi-generator-cli.jar generate -i petstore.yaml --generator-name spring -o mockservice -c java-config.json

- "openapi-generator-virtualan" project would developed on top of the  Open API Code generation project. This is supporting module that helps to Setup and Access all virtualized services.
                 - Creating the data-setup for the selected Service via UI - Look and feel similar to Swagger UI.
                 - Expose "Virtualized Service" as  service(OAS API). Utilized 
                                - Automated API Testing.
                                         

- Using cucumber setup the service Data in the Virtualized service to Setup data for the Resource-API Call would be tested
- Run  the API Calls  and Validate the response.
- Will  be creating some testcases ASAP using Cucumber with this deliverable.
- Certify the functional requirement in DEV Environment without Real Implemented services .
 and Etc..

jmini commented:
@elans3 sorry for the delay. As I wrote on Gitter with big contribution like this, it is sometime hard to find time to analyse them in details.

---

Is my understanding correct? This me trying to explain what your feature is about:

> “Service Virtualization Feature” (for the moment code proposed under `modules/openapi-generator-virtualan`) is a feature that can be added to a Spring Server to mock services without having to implement them.
> This is a sort of plugin for a spring server.

Feel free to tell if I got something wrong.

---

I think the “Service Virtualization” plugin does not belongs in this repository. My reasons:

* It will have an other lifecycle than `openapi-generator`
* It is not tighlty coupled to `openapi-generator`

This repository (`OpenAPITools/openapi-generator`) is about:
* one main module: `openapi-generator` containing the main logic
* and some additionals modules that add convenient ways to consume the main logic:
  - `openapi-generator-cli`: for CLI usage
  - `openapi-generator-maven-plugin`: for usage in a maven build
  - `openapi-generator-gradle-plugin`: for usage in a gradle build
  - `openapi-generator-online`: for usage of the codegen though a central server


I see “Service Virtualization Feature” as something comparable to:

* `io.swagger.core.v3:swagger-jaxrs2` a lib for a JaxRS Server that extend JaxRS with OpenAPI feature
* `inflector` a java server framwork to serve REST Services without using a Framework like Spring or JaxRS.

So we create a new project `OpenAPITools/virtualan` where we can do a first release (version `1.0.0` or whatever version makes sense for this first release).

We define @elans3 as committer on the new project.

---

Then my understanding is that there is a second aspect in this PR: for the spring generator, there should be a boolean option to generate a new project that using “Service Virtualization Feature”. This is implemented in this PR, but as top level config. This needs to be changed to be a spring only config.

When activated a spring project is created. This generated project embed the “Service Virtualization Feature”.

---

Does this make sense?

elans3 commented:
@jmini   It does make sense.   Correct!! It would be a plugin that makes the "Spring" generated code into a Virtualized service.  
**Part 1:**
May be i am thinking of include the following changes as part of 3.0.1 release.  
- openapitools-codegen code changes to generate the virtual service 
- Maven plugin changes

**Part 2**
We can move the virtualan into a separate project.  Thanks for your feedback. 
Let me know how would I proceed further.

@wing328  Share your views..

jmini commented:
I think **Part 2** needs to be made first, because you need the stuff to be released in order to be able to reference it in the spring template.

I think maybe the new project should be called: `OpenAPITools/spring-virtualan`. It might be clearer... Using `org.openapitools.virtualan` as package name is OK for me.

elans3 commented:
@jmini  I do agree part 2 needed first.  But I would not like to represent as spring-virtualan. Currently it is supported in the spring template but I have a vision it can be used on any template(the way it was developed). So just name as OpenAPITools/virtualan.  I am fine with the package name. Please let me know the timeline and how would I get the new project created? Thanks again for quick response.

jmini commented:
My understanding is that some members of the teams wants to give your feature a try... (as part of the  process to approve your project). So you might to wait a little bit.

In order to prepare "OpenAPITools/virtualan" you can just create it under your GitHub account "elans3/virtualan" (as you would create it under "OpenAPITools/virtualan", I mean with the `org.openapitools.virtualan` package name and the `openapitools` maven coordinates) moving it or using your stuff as initial commit will not be that complicated.

elans3 commented:
@jmini, @wing328  -  I have moved the Virtualan code base to my local repository.  I have added the implementation example along with this. This helps to setup quickly on anyone workspace.   Please review and let me know..
**Virtualan code:** 
 - https://github.com/elans3/openapi-virtualan

**Approach -1**  : Implementation sample code with pet store :
- https://github.com/elans3/openapi-virtualan-impl

**Approach -2** 
- You can add @ApiVirtual annotations in the method level. To make existing service as Mock/Virtualized. 
- Include the following dependency in your pom.xml
		<groupId>org.openapitools</groupId>
		<artifactId>openapi-virtualan</artifactId>
		<version>1.0.0</version>
- At present it supports only for Spring REST services only: 
- Deploy the code and then Navigate to the http://localhost:8080/virtualan-ui.html 

Thanks,
Elan

elans3 commented:
• Quick setup guide: 
**Pre-requisite:**
	- openapi-generator (With Virtualan version)
           https://github.com/elans3/openapi-generator
	- Virtualan code:
	https://github.com/elans3/openapi-virtualan

**Approach -1** : Implementation sample code with pet store :
	- https://github.com/elans3/openapi-virtualan-impl
 
**Approach -2**:
	- Spring REST Service should be Generated via openapi-generator. (As it uses the io.swagger.annotations.*  like ApiOperation and ApiResponses)
	- You can add @ApiVirtual annotations in the method level. To make existing service(Just that specific http action  as Mock/Virtualized).
	- Include the following dependency in your pom.xml (org.openapitools/openapi-virtualan)

• **Deploy and Start the service**:
	- It just supports only for Spring REST services as of now:
	- Run the following command:
		○ mvn -U -Dmaven.javadoc.skip=true clean install  
	- Run the following command
		○ java -jar virtualan-pet.jar 
• **Usage and Testing**:
	- Navigate to the http://localhost:<8080>/virtualan-ui.html to set up the Mock data.
	- UI is user friendly.. Don’t have any special documents..
	- Still need to get the test cases created.. Will be done soon. 
		○ Let me what all are the YAML should needs to be covered?
Hope this helps :)

wing328 commented on modules/openapi-generator-maven-plugin/README.md:
```diff
@@ -61,6 +61,7 @@ mvn clean compile
 - `generateSupportingFiles` - generate the supporting files (`true` by default)
 - `supportingFilesToGenerate` - A comma separated list of supporting files to generate.  All files is the default.
 - `skip` - skip code generation (`false` by default. Can also be set globally through the `codegen.skip` property)
+- `virtualService` - virtualized service code generation (`false` by default. Only available if `generator-name` is `spring`)
```
I would suggest not documenting the option here as it's a generator-specified option but not a top-level option.

jmini replied:
I think it is important not prepare stuff in advance... Put it at spring config for the beginning. If you implement this in an other java-server, then we will see what makes sense.

Global level is wrong in my opinion: it will be very confusing for people generating a ruby or a node server.

elans3 replied:
Make sense.. I had removed that from the top level.

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/config/CodegenConfigurator.java:
```diff
@@ -497,7 +498,7 @@ public ClientOptInput toClientOptInput() {
         checkAndSetAdditionalProperty(releaseNote, CodegenConstants.RELEASE_NOTE);
         checkAndSetAdditionalProperty(httpUserAgent, CodegenConstants.HTTP_USER_AGENT);
-
+        checkAndSetAdditionalProperty(virtualService,  CodegenConstants.VIRTUAL_SERVICE);
```
I suggest not making it a top-level option as it's specified to the spring generator only

wing328 commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-boot/pom.mustache:
```diff
@@ -134,6 +134,28 @@
+<!-- START Virtual Service API support -->
+		<dependency>
+			<groupId>org.openapitools</groupId>
+			<artifactId>openapi-virtualan</artifactId>
```
Is it correct to say that you also want to host this project under "org.openapitools"?

In other words, we'll need to move the project from https://github.com/elans3/openapi-virtualan to https://github.com/openapitools/openapi-virtualan, right?

jmini replied:
@wing328 : yes this was my personal suggestion. I think OpenAPITools organization can have other projects on board.

But this is something that needs to be validated by the team.

elans3 replied:
@jmini - I agree with you.. Just brought that just we had started with the discussion :)

elans3 commented:
@jmini  @wing328 : Follow-up,  what would be the plan for this PR.  Please share your thoughts.

elans3 commented:
Virtualan had released under io.virtualan.virtualan-plugin with 1.0.0. I had integrated the Open API Tools with Spring boot apps with Virtualan. Now it is available via https://oss.sonatype.org/content/repositories/releases/io/virtualan/virtualan-plugin/1.0.0 - the first version of virtualan..
I had updated my pull request(OpenAPITools/openapi-generator#196) with virtualan-plugin. Hope this should be good to proceed further.. Let me know anything else needed.

jmini commented:
Where is the repository holding the sources?

Why are you not hosting your released jar on maven central?

elans3 commented:
Yes it is available in the maven repo. https://repo.maven.apache.org/maven2/io/virtualan/virtualan-plugin/ 
Source Code: I am hosting under my own organisation.

wing328 commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-boot/pom.mustache:
```diff
@@ -134,6 +134,28 @@
+		    <version>1.5.9.RELEASE</version>
+		</dependency>
+    <!-- END Virtual Service API support -->
+{{/virtualService}}       
```
To make it consistent with the rest of the pom file, please use spaces instead of tabs

elans3 replied:
Removed the Tab spaces

elans3 commented on bin/springboot-virtualan-petstore-server.json:
```diff
@@ -0,0 +1,7 @@
+    "artifactId": "springboot-virtualan",
+    "library": "spring-boot",
+    "virtualService": true,
+	"modelPackage" : "org.openapitools.virtualan.model",
```
Removed the tab spaces

elans3 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConstants.java:
```diff
@@ -232,6 +232,9 @@
     public static final String GENERATE_MODEL_DOCS = "generateModelDocs";
 
+    public static final String VIRTUAL_SERVICE = "virtualService";
+    public static final String VIRTUAL_SERVICE_DOCS = "virtualServiceDocs";
```
Updated as Suggested

elans3 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:
```diff
@@ -113,6 +116,7 @@ public SpringCodegen() {
         cliOptions.add(CliOption.newBoolean(ASYNC, "use async Callable controllers", async));
         cliOptions.add(CliOption.newBoolean(REACTIVE, "wrap responses in Mono/Flux Reactor types (spring-boot only)", reactive));
         cliOptions.add(new CliOption(RESPONSE_WRAPPER, "wrap the responses in given type (Future,Callable,CompletableFuture,ListenableFuture,DeferredResult,HystrixCommand,RxObservable,RxSingle or fully qualified type)"));
+        cliOptions.add(CliOption.newBoolean(VIRTUAL_SERVICE, "Default generate without virtual service"));
```
Updated with Wiki reference.

wing328 commented:
@elans3 I ran `/bin/sh ./bin/springboot-virtualan-petstore-server.sh` and got the following update:

```diff
--- a/samples/server/petstore/springboot-virtualan/README.md
+++ b/samples/server/petstore/springboot-virtualan/README.md
@@ -1,29 +1,18 @@
 # OpenAPI generated server

-Spring Boot Virtual API Service using Virtulan: https://github.com/elan-venture/virtualan/wiki
+Spring Boot Server


 ## Overview
-This virtual service was generated by the [OpenAPI Generator](https://openapi-generator.tech) project.
-By using the [OpenAPI-Spec](https://openapis.org), you can easily generate a virtual service.
-This is an example of building a OpenAPI-enabled virtual service in Java using the SpringBoot framework.
+This server was generated by the [OpenAPI Generator](https://openapi-generator.tech) project.
+By using the [OpenAPI-Spec](https://openapis.org), you can easily generate a server stub.
+This is an example of building a OpenAPI-enabled server in Java using the SpringBoot framework.

-The underlying library integrating OpenAPI to SpringBoot is [springfox](https://github.com/springfox/springfox) and Virtualan (https://github.com/elan-venture/virtualan)
+The underlying library integrating OpenAPI to SpringBoot is [springfox](https://github.com/springfox/springfox)

-Build:
+Start your server as a simple java application

-  mvn clean install
+You can view the api documentation in swagger-ui by pointing to
+http://localhost:80/

-Run using standalone JAR:
-
-  java -jar target/virtualan-spring-controller-example-1.0.0.jar
-
-Invoke User REST service:
-
-  curl http://localhost/users
-
-Invoke Virtualan UI:
-
-Navigate to http://localhost/virtualan-ui.html.
-
-More details about the user interface refer- https://github.com/elan-venture/virtualan/wiki
\ No newline at end of file
+Change default port value in application.properties
\ No newline at end of file
```

Looks like all the Virtualan related information are gone. Did you see similar update when running the script locally?

wing328 commented:
@elans3 thanks for the quick fix. The change looks good to me.

elans3 commented:
@wing328.. Thank you!!!

elans3 commented:
Thank you!!!
